src/fs/backend/bootstrapping.cc

```diff
@@ -0,0 +1,579 @@
+        auto irt = taken_clusters.insert(std::move(nh));
+        if (!irt.inserted) {
+            mlogger.debug("^ Error: metadata and data log overlap by using the same cluster: {}", irt.node.value());
+            return METADATA_AND_DATA_LOG_OVERLAP;
+        }
+    }
+
+    // TODO: maybe check that writes do not overlap with other (still valid) writes on disk and do not use metadata_log clusters
+
+    std::deque<cluster_id_t> free_clusters;
+    for (auto cid : boost::irange(_available_clusters.beg, _available_clusters.end)) {
+        if (taken_clusters.find(cid) == taken_clusters.end()) {
+            free_clusters.emplace_back(cid);
+        }
+    }
+
+    mlogger.debug("free clusters: {}", free_clusters.size());
+    _shard._cluster_allocator = cluster_allocator(std::move(taken_clusters), std::move(free_clusters));
+
+    auto cid_opt = _shard._cluster_allocator.alloc();
+    if (!cid_opt) {
+        mlogger.debug("^ Error: could not allocate cluster for medium data log");
+        return NOSPACE;
+    }
+
+    _shard._medium_data_log_cw = _shard._medium_data_log_cw->virtual_constructor();
+    _shard._medium_data_log_cw->init(_shard._cluster_size, _shard._alignment,
+            cluster_id_to_offset(*cid_opt, _shard._cluster_size));
+    return SUCCESS;
+}
+
+void bootstrapping::initialize_shard_inode_allocator(fs_shard_id_t fs_shards_pool_size, fs_shard_id_t fs_shard_id) {
+    // Reset _inode_allocator
+    std::optional<inode_t> latest_allocated_inode;
+    if (!_shard._inodes.empty()) {
+        latest_allocated_inode =_shard._inodes.rbegin()->first;
+    }
+    _shard._inode_allocator = shard_inode_allocator(fs_shards_pool_size, fs_shard_id, latest_allocated_inode);
+}
+
+future<> bootstrapping::read_curr_cluster() {
+    disk_offset_t curr_cluster_disk_offset = cluster_id_to_offset(_curr_cluster.id, _shard._cluster_size);
+    mlogger.debug("Bootstrapping from cluster {}...", _curr_cluster.id);
+    return _shard._device.read(curr_cluster_disk_offset, _curr_cluster.content.get_write(),
+            _shard._cluster_size).then([this](size_t bytes_read) {
+        if (bytes_read != _shard._cluster_size) {
+            mlogger.debug("^ Error: partial read {} of {}", bytes_read, _shard._cluster_size);
+            return make_exception_future(failed_ml_cluster_read_exception());
+        }
+
+        mlogger.debug("Read cluster {}", _curr_cluster.id);
+        _curr_cluster.unparsed_content =
+            string_view(reinterpret_cast<const char*>(_curr_cluster.content.get()), _shard._cluster_size);
+        return now();
+    });
+}
+
+namespace {
+enum [[nodiscard]] rv_checkpoint_error {
+    NO_MEM,
+    BUFF_TOO_SMALL,
+    INVALID_ENTRY_TYPE,
+    INVALID_CHECKSUM
+};
+} // namespace
+
+// Returns checkpointed data or error
+static val_or_err<string_view, rv_checkpoint_error> read_and_validate_checkpoint(string_view& buff, disk_offset_t cluster_size) {
+    mlogger.debug("=========================");
```

Comment:
I know it's just debug logging, but avoid inserting arbitrary ornaments like this one - it's enough to just print the message below

Reply:
Ok, my intention here was to help readability while reading the debug log.